SU2_CFD/src/output/filewriter/CCGNSFileWriter.cpp

```diff
@@ -0,0 +1,310 @@
+#ifdef HAVE_CGNS
+  /*--- Open the CGNS file for writing.  ---*/
+  initializeMeshFile();
+
+  /*--- Write point coordinates. ---*/
+  WriteCoordinate(1, "CoordinateX");
+  WriteCoordinate(2, "CoordinateY");
+  if (nDim == 3) WriteCoordinate(3, "CoordinateZ");
+
+  /*--- Write mesh connectivity. ---*/
+  if (nDim == 2) {
+    WriteConnectivity(LINE, "Lines");
+    WriteConnectivity(TRIANGLE, "Triangles");
+    WriteConnectivity(QUADRILATERAL, "Quadrilaterals");
+  }
+  if (nDim == 3) {
+    WriteConnectivity(TRIANGLE, "Triangles");
+    WriteConnectivity(QUADRILATERAL, "Quadrilaterals");
+    WriteConnectivity(TETRAHEDRON, "Tetrahedra");
+    WriteConnectivity(PYRAMID, "Pyramids");
+    WriteConnectivity(PRISM, "Prisms");
+    WriteConnectivity(HEXAHEDRON, "Hexahedra");
+  }
+
+  /*--- Initialize and write fields. ---*/
+  InitializeFields();
+
+  const vector<string> fieldNames = dataSorter->GetFieldNames();
+  for (unsigned long i = nDim; i < fieldNames.size(); ++i) {
+    WriteField(i, fieldNames[i]);
+  }
+
+  /*--- Close the CGNS file. ---*/
+  if (rank == MASTER_NODE) CallCGNS(cg_close(cgnsFileID));
+
+#endif
+}
+
+#ifdef HAVE_CGNS
+void CCGNSFileWriter::initializeMeshFile() {
+  if (!dataSorter->GetConnectivitySorted()) {
+    SU2_MPI::Error("Connectivity must be sorted.", CURRENT_FUNCTION);
+  }
+
+  int nCell;
+
+  nLocalPoints = static_cast<int>(dataSorter->GetnPoints());
+  nDim = dataSorter->GetnDim();
+  GlobalElem = static_cast<cgsize_t>(dataSorter->GetnElemGlobal());
+  GlobalPoint = static_cast<cgsize_t>(dataSorter->GetnPointsGlobal());
+  cumulative = 0;
+
+  /*--- If surface file cell dimension is decreased. ---*/
+  nCell = nDim - isSurface;
+
+  if (rank == MASTER_NODE) {
+    /*--- Remove the previous file if present. ---*/
+    remove(fileName.c_str());
+
+    /*--- Create CGNS file and open in write mode. ---*/
+    CallCGNS(cg_open(fileName.c_str(), CG_MODE_WRITE, &cgnsFileID));
+
+    /*--- Create Base. ---*/
+    CallCGNS(cg_base_write(cgnsFileID, "Base", nCell, nDim, &cgnsBase));
+
+    /*--- Create Zone. ---*/
+    vector<cgsize_t> zoneData(3);
+
+    zoneData[0] = GlobalPoint;
+    zoneData[1] = GlobalElem;
+    zoneData[2] = 0;
+
+    CallCGNS(cg_zone_write(cgnsFileID, cgnsBase, "Zone", zoneData.data(), Unstructured, &cgnsZone));
+  }
+}
+
+void CCGNSFileWriter::WriteCoordinate(int CoordinateNumber, const string& CoordinateName) {
+  cgsize_t nodeBegin;
+  cgsize_t nodeEnd;
+  vector<float> sendBufferFloat, recvBufferFloat;
+
+  /*--- Create send buffer. ---*/
+  sendBufferFloat.resize(nLocalPoints);
+
+  unsigned short iCoord = CoordinateNumber - 1;
+  for (int iPoint = 0; iPoint < nLocalPoints; iPoint++) {
+    sendBufferFloat[iPoint] = static_cast<float>(dataSorter->GetData(iCoord, iPoint));
+  }
+
+#ifdef HAVE_MPI
+  SU2_MPI::Request sendReq;
+  if (rank != MASTER_NODE)
+    SU2_MPI::Isend(sendBufferFloat.data(), nLocalPoints, MPI_FLOAT, MASTER_NODE, 0, SU2_MPI::GetComm(), &sendReq);
```

Comment:
If you don't need non-blocking use the regular Send
```suggestion
  if (rank != MASTER_NODE) {
    SU2_MPI::Send(sendBufferFloat.data(), nLocalPoints, MPI_FLOAT, MASTER_NODE, 0, SU2_MPI::GetComm());
    return;
  }
```